Kernel/Devices/SB16.cpp

```diff
@@ -0,0 +1,183 @@
+#include "SB16.h"
+#include "IO.h"
+#include <Kernel/VM/MemoryManager.h>
+
+const int DSP_READ = 0x22A;
+const int DSP_WRITE = 0x22C;
+const int DSP_STATUS = 0x22E;
+const int DSP_R_ACK = 0x22F;
+
+/* Write a value to the DSP write register */
+void SB16::dsp_write(uint8_t value)
+{
+    while (IO::in8(DSP_WRITE) & 0x80) {
+        ;
+    }
+    IO::out8(DSP_WRITE, value);
+}
+
+/* Reads the value of the DSP read register */
+uint8_t SB16::dsp_read()
+{
+    while (!(IO::in8(DSP_STATUS) & 0x80)) {
+        ;
+    }
+    return IO::in8(DSP_READ);
+}
+
+/* Changes the sample rate of sound output */
+void SB16::set_sample_rate(uint16_t hz)
+{
+    dsp_write(0x41); // output
+    dsp_write((uint8_t)(hz >> 8));
+    dsp_write((uint8_t)hz);
+    dsp_write(0x42); // input
+    dsp_write((uint8_t)(hz >> 8));
+    dsp_write((uint8_t)hz);
+}
+
+static SB16* s_the;
+
+SB16::SB16()
+    : IRQHandler(5)
+    , CharacterDevice(42, 42) // ### ?
+{
+    s_the = this;
+    initialize();
+}
+
+SB16::~SB16()
+{
+}
+
+SB16& SB16::the()
+{
+    return *s_the;
+}
+
+void SB16::handle_irq()
+{
+    m_interrupted = true;
+
+    // Stop sound output ready for the next block.
+    dsp_write(0xd0);
+
+    IO::in8(DSP_STATUS); // 8 bit interrupt
+    if (m_vmaj >= 4)
+        IO::in8(DSP_R_ACK); // 16 bit interrupt
+}
+
+void SB16::initialize()
+{
+    int32_t i;
+    IO::out8(0x226, 1);
+    for (i = 0; i < 100000; i++) {
+        ; // ### wait
+    }
+    IO::out8(0x226, 0);
+
+    auto data = dsp_read();
+    if (data != 0xaa) {
+        kprintf("SB16: sb not ready");
+        return;
+    }
+
+    // Get the version info
+    dsp_write(0xe1);
+    m_vmaj = dsp_read();
+    auto vmin = dsp_read();
+
+    kprintf("SB16: found version %d.%d\n", m_vmaj, vmin);
+    enable_irq();
+}
+
+bool SB16::can_read(FileDescription&) const
+{
+    return false;
+}
+
+ssize_t SB16::read(FileDescription&, u8*, ssize_t)
+{
+    return 0;
+}
+
+void SB16::dma_start(uint32_t length)
+{
+    const auto addr = m_dma_buffer_page->paddr().get();
+    const u8 channel = 1;
+    const u8 mode = 0;
+
+    // Disable the DMA channel
+    const int BIT_4 = 0x04;
```

Comment:
`BIT_4` seems like an odd name for the value `0x04`. Does it mean something in the context of the SB16?

Reply:
https://homepages.cae.wisc.edu/~brodskye/sb16doc/sb16doc.html has been the best source I've been working from. This was a leftover from trial-and-error, I've fixed it accordingly.